examples/petsc/qfunctions/bps/bp1sphere.h

```diff
@@ -0,0 +1,195 @@
+/// libCEED QFunctions for mass operator example for a scalar field on the sphere using PETSc
+
+#ifndef __CUDACC__
+#  include <math.h>
+#endif
+
+// *****************************************************************************
+// This QFunction sets up the geometric factors required for integration and
+//   coordinate transformations when reference coordinates have a different
+//   dimension than the one of physical coordinates
+//
+// Reference (parent) 2D coordinates: X \in [-1, 1]^2
+//
+// Global 3D physical coordinates given by the mesh: xx \in [-R, R]^3
+//   with R radius of the sphere
+//
+// Local 3D physical coordinates on the 2D manifold: x \in [-l, l]^3
+//   with l half edge of the cube inscribed in the sphere
+//
+// Change of coordinates matrix computed by the library:
+//   (physical 3D coords relative to reference 2D coords)
+//   dxx_j/dX_i (indicial notation) [3 * 2]
+//
+// Change of coordinates x (on the 2D manifold) relative to xx (phyisical 3D):
+//   dx_i/dxx_j (indicial notation) [3 * 3]
+//
+// Change of coordinates x (on the 2D manifold) relative to X (reference 2D):
+//   (by chain rule)
+//   dx_i/dX_j [3 * 2] = dx_i/dxx_k [3 * 3] * dxx_k/dX_j [3 * 2]
+//
+// modJ is given by the magnitude of the cross product of the columns of dx_i/dX_j
+//
+// The quadrature data is stored in the array qdata.
+//
+// We require the determinant of the Jacobian to properly compute integrals of
+//   the form: int( u v )
+//
+// Qdata: modJ * w
+//
+// *****************************************************************************
+
+// *****************************************************************************
+CEED_QFUNCTION(SetupMassGeo)(void *ctx, const CeedInt Q,
+                             const CeedScalar *const *in,
+                             CeedScalar *const *out) {
+  // Inputs
+  const CeedScalar *X = in[0], *J = in[1], *w = in[2];
+  // Outputs
+  CeedScalar *qdata = out[0];
+
+  // Quadrature Point Loop
+  CeedPragmaSIMD
+  for (CeedInt i=0; i<Q; i++) {
+    // Read global Cartesian coordinates
+    const CeedScalar xx[3][1] = {{X[i+0*Q]},
+                                 {X[i+1*Q]},
+                                 {X[i+2*Q]}
+                                };
+
+    // Read dxxdX Jacobian entries, stored as
+    // 0 3
+    // 1 4
+    // 2 5
+    const CeedScalar dxxdX[3][2] = {{J[i+Q*0],
+                                     J[i+Q*3]},
+                                    {J[i+Q*1],
+                                     J[i+Q*4]},
+                                    {J[i+Q*2],
+                                     J[i+Q*5]}
+                                   };
+
+    // Setup
+    // x = xx (xx^T xx)^{-1/2}
+    // dx/dxx = I (xx^T xx)^{-1/2} - xx xx^T (xx^T xx)^{-3/2}
+    const CeedScalar modxxsq = xx[0][0]*xx[0][0]+xx[1][0]*xx[1][0]+xx[2][0]*xx[2][0];
+    CeedScalar xxsq[3][3];
+    for (int j=0; j<3; j++)
+      for (int k=0; k<3; k++) {
+        xxsq[j][k] = 0;
+        for (int l=0; l<1; l++)
```

Comment:
Why are there a number of vectors that are created and indexed as 3x1 matrices? Seems needlessly complicated.

Reply:
Sorry I don't know why I had started working that way. Simplified now.